File: JavaPerformanceTest/src/akshay/javaperformance/StoreString.java
package akshay.javaperformance;

import java.util.ArrayList;

public class StoreString {

	public static int binarySearch(ArrayList<RandomString> list, int first, int upto, String key) {
		if (first < upto) {
			int mid = first + (upto - first) / 2;  // Compute mid point.
			if (Integer.parseInt(key) < list.get(mid).getNum()) {
				return binarySearch(list, first, mid, key);		            
			} else if (Integer.parseInt(key) > list.get(mid).getNum()) {
				return binarySearch(list, mid+1, upto , key);		            
			} else {
				return mid;   // Found it.
			}
		}
		return -(first + 1);  // Failed to find key
	}

	public static ArrayList<RandomString> quicksort(ArrayList<RandomString> list) {
		if (list.size() <= 1)
			return list;
		int pivot = list.size() / 2;
		ArrayList<RandomString> lesser = new ArrayList<RandomString>();
		ArrayList<RandomString> greater = new ArrayList<RandomString>();
		int sameAsPivot = 0;
		for (RandomString number : list) {
			if (number.getNum() > list.get(pivot).getNum())
				greater.add(number);
			else if (number.getNum() < list.get(pivot).getNum())
				lesser.add(number);
			else
				sameAsPivot++;
		}
		lesser = quicksort(lesser);
		for (int i = 0; i < sameAsPivot; i++)
			lesser.add(list.get(pivot));
		greater = quicksort(greater);
		ArrayList<RandomString> sorted = new ArrayList<RandomString>();
		for (RandomString number : lesser)
			sorted.add(number);
		for (RandomString number: greater)
			sorted.add(number);
		return sorted;
	}
}

File: ScalaPerformanceTest/bin/akshay/scalaperformance/StoreString.scala
package akshay.scalaperformance

object StoreString {
 
	def quicksort(l: List[RandomString]): List[RandomString] = {
			l match {
			case List() => l
			case _ =>  quicksort(for(x <- l.tail if x.num < l.head.num) yield x) ::: List(l.head) ::: quicksort(for(x <- l.tail if x.num >= l.head.num) yield x)
			}
	}

	def binarySearch(s: String, list: List[RandomString],low: Int, high: Int): Int = {
			(high-low) match {
			case _ if (high-low) > 0 => {
				java.lang.Integer.parseInt(s) match {
				case _ if java.lang.Integer.parseInt(s) < list(low + (high-low)/2).num => binarySearch(s, list, low, low + (high-low)/2)
				case _ if java.lang.Integer.parseInt(s) > list(low + (high-low)/2).num => binarySearch(s, list, (low + (high-low)/2)+1, high)
				case _  => (low + (high-low)/2)+1
				}
			}
			case _ if (high-low) <= 0 => -1
			}
	}
}

